botocore/docs/bcdoc/docstringparser.py

```diff
@@ -22,30 +21,175 @@ class DocStringParser(html_parser.HTMLParser):
     """
 
     def __init__(self, doc):
-        html_parser.HTMLParser.__init__(self)
+        six.moves.html_parser.HTMLParser.__init__(self)
         self.doc = doc
-        self.unhandled_tags = []
+        self.tree = HTMLTree(self.doc)
 
     def handle_starttag(self, tag, attrs):
-        handler_name = 'start_%s' % tag
-        if hasattr(self.doc.style, handler_name):
-            getattr(self.doc.style, handler_name)(attrs)
-        else:
-            self.unhandled_tags.append(tag)
+        self.tree.add_tag(tag, attrs=attrs)
 
     def handle_endtag(self, tag):
-        handler_name = 'end_%s' % tag
-        if hasattr(self.doc.style, handler_name):
-            getattr(self.doc.style, handler_name)()
+        self.tree.add_tag(tag, is_start=False)
 
     def handle_data(self, data):
-        if data.isspace():
-            data = ' '
+        self.tree.add_data(data)
+
+    def close(self):
+        # HTMLParser is an old style class, so the super() method will not work.
+        six.moves.html_parser.HTMLParser.close(self)
+        self.tree.write()
```

Comment:
I can imagine a few ways to write as we go, but I think that would make the code much more complicated. We have no forward context if we do that, and no way to go back on previous writes. We would need to either give every node `li` specific state, or do a reverse tree traversal to find out if it is a child of an `li` and then back down to see if it is the first instance of text. I agree it is unfortunate that we have to buffer the whole thing, but I think for the very specific case of doc strings, it is fine. We know that the doc strings will never be more than a few paragraphs, whereas if we buffered ALL the docs (even for a single service) it could be a massive memory footprint.